src/main/process-pool.ts

```diff
@@ -0,0 +1,70 @@
+import { ChildProcess, fork } from 'child_process';
+import * as os from 'os';
+import { resolve as pathResolve } from 'path';
+import { removeDebugArgs } from './util';
+
+interface IMochaProcess extends ChildProcess {
+  destroy: () => void;
+}
+
+export default class ProcessPool {
+  private maxParallel: number;
+  private waitingList: Array<(process: IMochaProcess) => void> = [];
+  private unusedProcesses: IMochaProcess[] = [];
+  private processes: IMochaProcess[] = [];
+
+  constructor() {
+    this.maxParallel = os.cpus().length;
+  }
+
+  setMaxParallel(n) {
+    this.maxParallel = n;
+  }
+
+  async getOrCreate(isTypescriptRunMode): Promise<IMochaProcess> {
+    const extension = isTypescriptRunMode ? 'ts' : 'js';
+    const runnerPath = pathResolve(__dirname, `../subprocess/runner.${extension}`);
+
+    const lastUnusedProcess = this.unusedProcesses.pop();
+    if (lastUnusedProcess) {
+      return lastUnusedProcess;
+    }
+
+    if (this.processes.length >= this.maxParallel) {
+      const process: IMochaProcess = await new Promise<IMochaProcess>((resolve) => {
+        this.waitingList.push((proc: IMochaProcess) => {
+          resolve(proc);
+        });
+      });
+      return process;
+    }
+    return this.create(runnerPath, {
+      // otherwise `--inspect-brk` and other params will be passed to subprocess
+      execArgv: process.execArgv.filter(removeDebugArgs),
+      stdio: ['ipc'],
+    });
+  }
+
+  create(runnerPath, opt) {
+    const process = fork(runnerPath, [], opt) as IMochaProcess;
+
+    this.processes.push(process);
+
+    process.destroy = () => {
+      const nextOnWaitingList = this.waitingList.pop();
+      if (nextOnWaitingList) {
+        nextOnWaitingList(process);
+      } else {
+        this.unusedProcesses.push(process);
+      }
+    };
+
+    return process;
+  }
```

Comment:
One more thing to consider here is the mocha `--exit` option. It looks like you call this method right after the `end` event is got from the subprocess. But in fact, that's not the time when the process is finished: it can have an asynchronous logic which can for instance stop listening for any kind of events and the process can exit with a non-zero exit code in it. In this case, although the test didn't have any errors mocha will show it as failed. So please take a closer look at #155 and maybe #169 comments. Maybe the thing we should do is to use process pool only when `mocha` is launched with an `--exit` option.

Reply:
What do we do then?

Reply:
That's a question to you. It looks to me like we can re-use processes only when `--exit` option is specified otherwise we break the `mocha` behaviour.

Reply:
@fabiosantoscode ^ any ideas? If no, let's enable this behaviour for `--exit`-run tests and add a test for that.